EIPS/eip-9999.md

```diff
@@ -0,0 +1,150 @@
+
+#### Transaction Processing
+
+```python
+def process_transaction(env: vm.Environment, tx: Transaction) -> Tuple[Uint, Tuple[Log, ...], Optional[Exception]]:
+    preaccessed_addresses = set()
+    preaccessed_storage_keys = set()
+    
+    # Add block-level pre-accessed slots
+    preaccessed_addresses.add(env.block_level_accessed_addresses)
+    preaccessed_storage_keys.add(env.block_level_accessed_storage_keys)
+    
+    # Handle access lists from transaction
+    ...
+```
+
+This adds the block-level-accessed addresses and storage keys to the preaccessed addresses and storage keys.
+As a result, from the perspective of a transaction, block-level accessed addresses and storage keys are treated the same as precompiles or the coinbase address.
+
+```python
+def process_message_call(message: Message, env: Environment) -> MessageCallOutput:
+    return MessageCallOutput(
+        # ... other fields ...
+        accessed_addresses=evm.accessed_addresses,
+        accessed_storage_keys=evm.accessed_storage_keys
+    )
+```
+
+The message call processing tracks accessed addresses and storage keys during execution, which are then propagated back up to the transaction level and ultimately to the block level.
+
+## Rationale
+
+The proposal builds on several key observations:
+
+1. **Caching Efficiency**: Modern Ethereum clients already implement sophisticated caching mechanisms at the block level. Extending address and storage key warming to match this caching behavior better aligns gas costs with actual computational costs.
+
+2. **Backward Compatibility**: The worst-case scenario for any transaction remains unchanged - it will never pay more than the current cold access cost for its first access to a slot.
+
+3. **First Access Warms For All**: Under the proposed mechanism, the first transaction in a block that accesses and warms multiple addresses or storage slots bears the entire cost of warming, even if subsequent transactions benefit from those already-warmed slots without incurring additional costs. This approach is deemed acceptable because early execution within a block—where warming typically occurs—is generally occupied for transactions placed by professional builders aiming for top-of-block execution. Since the resulting cost discrepancy is relatively minor, a more complex cost-sharing model is avoided in favor of maintaining simplicity.
+An alternative to the one-warms-for-all mechanism involves distributing the warming costs across all accesses within a block. In this approach, each transaction must at least be able to pay the cold access cost. Once all transactions in the block are executed, the total warming costs are evenly distributed, and transaction originators are refunded accordingly.
+
+4. **Alternative Block Warming Windows**: Instead of applying warming at the block level, more advanced multi-block warming approaches can be considered. Potential options include:
```

Comment:
I think first moving to "block-warm" state instead of "tx-warm" state is preferred. But this is certainly something to think about. The main complexity problem I see with this idea is that this ring buffer needs to be saved somewhere, and it needs to be part of the consensus, such that clients once shut down and starting up again can lookup whether accounts/storage are warm or not.

Reply:
Yeah the ring buffer adds some complexity for sure.
@pipermerriam also had a comment regarding additional complexity for statelessness [here](https://ethresear.ch/t/block-level-warming/21452/2?u=nero_eth).

Reply:
I don't see any problems with this as long as it only remains single-block.